bpo-22807: Expose platform UUID generation safety information.

warsaw opened the pull request:
Add an enum `SafeUUID` which describes whether the platform has generated a UUID in a safe manner (i.e. ensuring that no two processes can generate the same UUID).  Also add `UUID.is_safe` attribute which relays the information from the underlying platform.

warsaw committed:
bpo-22807: Expose platform UUID generation safety information.

warsaw committed:
bpo-22807: Expose platform UUID generation safety information.

warsaw committed:
pragma-fy some lines that are okay not to get covered.

warsaw committed:
pragma-fy some lines that are okay not to get covered.

Carreau commented on Lib/uuid.py:
```diff
@@ -182,6 +196,7 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
             int &= ~(0xf000 << 64)
             int |= version << 76
         self.__dict__['int'] = int
+        self.__dict__['is_safe'] = is_safe
```
Do you want to assert the type of `is_safe` ? There seem to extensive checking of the rest of the parameters.
Not that I imagine people creating UUID object in their own code, but `is_safe` feel also a lot like a Boolean.

warsaw replied:
I'm not sure it's worth type checking `is_safe`.  The other ones are checked to ensure sane, compliant UUIDs, but `is_safe` is just a flag with some extra information (which may or may not be helpful to an application).  If someone wants to manually instantiate UUIDs with bogus `is_safe` flags, oh well.

Carreau commented on Lib/uuid.py:
```diff
@@ -104,7 +114,7 @@ class UUID(object):
 
     def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
-                       int=None, version=None):
+                       int=None, version=None, is_safe=SafeUUID.unknown):
```
> I changed my mind on whether this should affect older versions of Python.

If so, you can also consider making it kwarg-only. As you are already using is as a kwarg. Not sure if that's useful, or worth it. It's may also be a question of taste, and leaving it as is seem perfectly fine.

warsaw replied:
Thanks for the review!  I did think about that, and it might indeed make sense, but it's a matter of style in this case, and I wasn't sure I wanted to introduce that to the API.  But on thinking about it again now that you bring it up, I does make sense.

warsaw committed:
Make is_safe a keyword argument.

warsaw committed:
Make is_safe a keyword argument.

warsaw committed:
Oops, fix indentation.

warsaw committed:
Oops, fix indentation.

Carreau reviewed (approved):
> Thanks for the review!

My pleasure !

> I did think about that, and it might indeed make sense, but it's a matter of style in this case, and I wasn't sure I wanted to introduce that to the API. But on thinking about it again now that you bring it up, I does make sense.

Happy to be of some use :-)

berkerpeksag commented on Doc/library/uuid.rst:
```diff
@@ -19,8 +19,30 @@ If all you want is a unique ID, you should probably call :func:`uuid1` or
-.. class:: UUID(hex=None, bytes=None, bytes_le=None, fields=None, int=None, version=None)
+.. class:: SafeUUID
+
+.. attribute:: SafeUUID.safe
```
I know this is the current style in uuid.rst, but it might be better use the following style in new documentation:

```rst
.. class:: SafeUUID

   .. versionadded:: 3.7

   .. attribute:: safe
      The UUID was generated by the platform in a multiprocessing-safe way.

   ...
```

warsaw replied:
Yeah, that looks nice, thanks.  The one problem is that `SafeUUID` still gets labeled as a class in the HTML.  Okay, technically it *is* a class, but it would be nice to have an `enum::` markup.  I grep'd around in the source and couldn't find any such examples, so I don't know if we support that yet.

berkerpeksag replied:
You're right. We don't have an enum directive now, but it should be easy to add one. You can add me to nosy list if you create an issue on bugs.p.o :)

berkerpeksag commented on Lib/test/test_uuid.py:
```diff
@@ -340,6 +340,46 @@ def test_uuid1(self):
         equal(((u.clock_seq_hi_variant & 0x3f) << 8) |
                          u.clock_seq_low, 0x3fff)
 
+    @unittest.skipUnless(importable('ctypes'), 'requires ctypes')
```
Nitpick: I'd flip the order of the decorators.

warsaw replied:
+1

berkerpeksag commented on Lib/test/test_uuid.py:
```diff
@@ -340,6 +340,46 @@ def test_uuid1(self):
+        # unknown (unless I suppose the platform is buggy).
+        self.assertNotEqual(u.is_safe, uuid.SafeUUID.unknown)
+
+    @unittest.skipUnless(importable('ctypes'), 'requires ctypes')
```
Just a suggestion: Using `self.subTest()` would make the following four tests much shorter and IMO easier to understand.

warsaw replied:
I've played a bit with `subTest()` and I'm not entirely thrilled with it.  AFAICT, subtests only print output when they fail, not when they succeed.  In verbose mode, I'd want to see all 4 tests printed individually.  (That's one of the things I'd like to eventually fix.)  For now, I think I'll leave them as separate tests.

orsenthil reviewed (approved):
LGTM.

berkerpeksag commented on Doc/library/uuid.rst:
```diff
@@ -19,8 +19,30 @@ If all you want is a unique ID, you should probably call :func:`uuid1` or
+
+.. versionadded:: 3.7
+
+.. class:: UUID(hex=None, bytes=None, bytes_le=None, fields=None, int=None, version=None, *, is_safe=SafeUUID.unknown)
```
Currently, the body of `UUID()` documentation briefly describes all its arguments so I think we can just copy is_safe documentation in `Lib/uuid.py`

warsaw replied:
+1

berkerpeksag reviewed (approved):
I just left some comments, but the PR looks good to me.

ncoghlan committed:
Make devguide link in README more prominent (#145)

ncoghlan committed:
bpo-29571: Use correct locale encoding in test_re (#149)

``local.getlocale(locale.LC_CTYPE)`` and
``locale.getpreferredencoding(False)`` may give different answers
in some cases (such as the ``en_IN`` locale).

``re.LOCALE`` uses the latter, so update the test case to match.

warsaw committed:
Merge branch 'bpo-22807' of github.com:warsaw/cpython into bpo-22807

warsaw committed:
Respond to review
